File: Javascript/Algorithms/257.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
let get_paths = (node, prev, result) => {
    let {val, left, right} = node;
    // build path
    if (prev == "")
        prev = `${val}`
    else 
        prev = `${prev}->${val}`;
    
    if (left)
        get_paths(left, prev, result);
    if (right)
        get_paths(right, prev, result);
    // Add path to result when we reach a leaf.
    if (!left && !right){
        result.push(prev);
    }
    return result;
}

let binaryTreePaths = root => {
    if (root){
        return get_paths(root, "", []);
    }
    return [];
};

File: Python/Algorithms/257.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root):
        def get_paths(node, prev):
            # build the result partially.
            sval = repr(node.val)
            res = "->".join([prev, sval]) if prev else sval

            # until we have left or right, recurse.
            left, right = node.left, node.right
            if left:
                yield from get_paths(left, res)
            if right:
                yield from get_paths(right, res)
            # yield back when we reach a leaf.
            if not left and not right:
                yield res

        if root:
            return [*get_paths(root, "")]
        return []
